### PR TITLE
#162685661 Implement a Password Reset functionality

### PR DESCRIPTION
**Pull request by moseskamira**
####What does this PR do?
This PR delivers a feature that enables a user to reset a password and receive a confirmation email before reset

#### Description of Task to be completed?
A User shall be in position to reset a password using this feature when completed

#### How should this be manually tested?

- Make sure you have one VALID email addresses on gmail with their corresponding password.
  Signup onto the APP using this  email address that you shall use to login and the request a password reset 

- Create a .env file on the root directory and add the data below, substituting each variable as appropriately described.
EMAIL_PORT=587
EMAIL_HOST_USER = gauthorshaven@gmail.com
EMAIL_HOST_PASSWORD = guardiansauthors
 

- Now run the url:  http://localhost:8000/api/password-reset/
- Fill in the email address you used to register as a user and click Reset my password button.
Use Format:
{
    "user": {
        "email": "yourvalidemail@gmail.com."
    }
}

- Follow the prompt and check your email address for the reset password message.
#### What are the relevant pivotal tracker stories?
#162685661

### CODE REVIEW

**Comment by archbishop12** [comment]
@moseskamira Coverage has dropped, could you fix that.

File: authors/settings.py [hunk 1]
```diff
@@ -158,3 +158,9 @@
     #     'authors.apps.authentication.backends.JWTAuthentication',
     # ),
 }
+EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
+EMAIL_HOST = 'smtp.gmail.com'
+EMAIL_PORT = config('EMAIL_PORT')
+EMAIL_USE_TLS = True
+EMAIL_HOST_USER = config('EMAIL_HOST_USER')
+EMAIL_HOST_PASSWORD = config('EMAIL_HOST_PASSWORD')
```
Comment (archbishop12) [inline authors/settings.py:166]: @moseskamira Could you create an email for the guardians that will be used for this. e.g gauthorshaven@gmail.com

File: authors/apps/authentication/views.py [hunk 2]
```diff
@@ -74,3 +85,79 @@ def update(self, request, *args, **kwargs):
...
+        user = request.data.get('user', {})
+        serializer = self.serializer_class(data=user)
+        serializer.is_valid(raise_exception=True)
+        
+        user = User.objects.filter(email=user['email']).first()
+        token = default_token_generator.make_token(user)
+        uid = urlsafe_base64_encode(force_bytes(user.username)).decode('utf-8')
+        domain = 'http://{}'.format(get_current_site(request))
+        route = 'api/password-reset-confirm/'
+        url = '{}/{}{}-{}'.format(domain, route, token, uid)
```
Comment (archbishop12) [inline authors/apps/authentication/views.py:106]: @moseskamira Could you create a separate method for the sending the email.

**Comment by archbishop12** [comment]
@moseskamira Please run the ```pep8``` on the files you edited and ensure that you no lines are flagged.

File: authors/apps/authentication/views.py [hunk 3]
```diff
@@ -74,3 +85,79 @@ def update(self, request, *args, **kwargs):
...
+
+        if serializer_data['password'] == serializer_data['confirm_password']:
+            serializer = self.serializer_class(
+                request.data, data=serializer_data, partial=True
+            )
+            serializer.is_valid(raise_exception=True)
+            new_password = make_password(serializer_data['confirm_password'])
+            User.objects.filter(username=user).update(password=new_password)
+
+            return Response({
```
Comment (fahadmak) [inline authors/apps/authentication/views.py:142]: Please add tests to cover this

File: authors/apps/authentication/views.py [hunk 4]
```diff
@@ -74,3 +85,79 @@ def update(self, request, *args, **kwargs):
...
+                "message": {
+                    "detail": [
+                        "Password Has Been Successfully Reset, Please Login "
+                    ]
+                }
+            }, status=status.HTTP_200_OK)
+        return Response({
```
Comment (fahadmak) [inline authors/apps/authentication/views.py:149]: please also add tests to cover this or refer to coveralls to check more edge cases that  have not been covered

File: authors/apps/authentication/tests/test_views.py [hunk 5]
```diff
@@ -44,3 +46,79 @@ def test_edit_user_credentials(self):
...
+        self.assertEqual(
+            json.loads(res.content),
+            {
+                'errors': {
+                    'email': [
+                        'This field may not be blank.'
+                    ]
+                }
+            }
+        )
```
Comment (fahadmak) [inline authors/apps/authentication/tests/test_views.py]: Always add a new line at the end of file

Comment (fahadmak) [reply authors/apps/authentication/tests/test_views.py]: Try using pep8

**Review by fahadmak — CHANGES_REQUESTED** [review]

**Comment by BarnaTB** [comment]
Test coverage has dropped just as @archbishop12 stated so you could check on that.

File: authors/apps/authentication/views.py (hunk 3 above)
Comment (VivianDoreen) [reply authors/apps/authentication/views.py]: The functionality is working successfully, great work. Though Improve your coverage by writing more tests

**Comment by VivianDoreen** [comment]
The functionality is working successfully, great work. Though Improve your coverage by writing more tests

File: authors/settings.py (hunk 1 above)
Comment (moseskamira) [reply authors/settings.py:166]: created and PR Information editted accordingly

**Review by BarnaTB — APPROVED** [review]
The feature runs properly but it'd be nice if you made the PT board story link and actual link and not plain text based on what we see right now. Great nonetheless!

**Review by fahadmak — APPROVED** [review]

**Review by BarnaTB — APPROVED** [review]
The feature runs properly but it'd be nice if you made the PT board story link and actual link and not plain text based on what we see right now. Great nonetheless!

File: authors/apps/authentication/serializers.py [hunk 6]
```diff
@@ -154,3 +154,31 @@ def update(self, instance, validated_data):
...
+            )
+        return email
+
+
+class ResetPasswordConfirmSerializer(serializers.Serializer):
+    """
+    Serializer Class For Confirming Change Of Password
+    """
+    password = serializers.CharField(max_length=128, min_length=8)
+    confirm_password = serializers.CharField(max_length=128, min_length=8)
```
Comment (archbishop12) [inline authors/apps/authentication/serializers.py:184]: @moseskamira Please add the a line at the end of this file.

File: authors/apps/authentication/tests/test_views.py [hunk 7]
```diff
@@ -44,3 +46,151 @@ def test_edit_user_credentials(self):
         username = user_update['user']['username']
         self.assertEqual(response.data['username'], username)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+
+class TestResetPassword(APITestCase):
+    def setUp(self):
```
Comment (archbishop12) [inline authors/apps/authentication/tests/test_views.py:52]: @moseskamira Could you sync with @fahadmak and use some of the data in `data.py` or any data you want there tat could be missing.

File: authors/apps/authentication/views.py [hunk 8]
```diff
@@ -74,3 +84,83 @@ def update(self, request, *args, **kwargs):
...
+
+class ResetPasswordAPIView(GenericAPIView):
+    permission_classes = (AllowAny,)
+    serializer_class = ResetPasswordSerializer
+
+    @classmethod
+    def decode_id(self, uid):
+        username = urlsafe_base64_decode(uid).decode('utf-8')
+        return username
+        # print(username)
```
Comment (archbishop12) [inline authors/apps/authentication/views.py]: @moseskamira Remove commented out line of code.

File: authors/apps/authentication/views.py [hunk 9]
```diff
@@ -74,3 +84,83 @@ def update(self, request, *args, **kwargs):
...
+
+
+class ResetPasswordConfirmAPIView(RetrieveUpdateAPIView):
+    permission_classes = (AllowAny,)
+    serializer_class = ResetPasswordConfirmSerializer
+
+    def get_object(self):
+        return {"password": "", "confirm_password": ""}
+
+    def update(self, request, **kwargs):
```
Comment (archbishop12) [inline authors/apps/authentication/views.py:130]: @moseskamira Please reduce the size of this method , you could remove the first try and expect block and put it in a separate method called `validate_slug` which can be called here..

File: authors/apps/authentication/views.py [hunk 10]
```diff
@@ -74,3 +84,85 @@ def update(self, request, *args, **kwargs):
...
+            }, status=status.HTTP_200_OK)
+        return Response({
+            "errors": {
+                "body": [
+                    "Password Mismatch, Please Re-enter Password"
+                ]
+            }
+        }, status=status.HTTP_400_BAD_REQUEST)
+
+    def validate_slug(self, **kwargs):
```
Comment (archbishop12) [inline authors/apps/authentication/views.py]: @moseskamira You created the method but you are not using it anywhere????

File: authors/settings.py [hunk 11]
```diff
@@ -158,3 +158,9 @@
     #     'authors.apps.authentication.backends.JWTAuthentication',
     # ),
 }
+EMAIL_BACKEND = 'django.core.mail.backends.smtp.EmailBackend'
+EMAIL_HOST = 'smtp.gmail.com'
+EMAIL_PORT = config('EMAIL_PORT')
+EMAIL_USE_TLS = True
+EMAIL_HOST_USER = 'gauthorshaven@gmail.com'
```
Comment (archbishop12) [inline authors/settings.py]: @moseskamira These should be added through the env

File: authors/settings.py [hunk 12]
```diff
@@ -158,3 +158,9 @@
...
+EMAIL_HOST_PASSWORD = 'guardiansauthors'
```
Comment (archbishop12) [inline authors/settings.py]: @moseskamira Even this, Sensitive information e.g `passwords` should not be added directly to the codebase.

**Review by VivianDoreen — APPROVED** [review]

**Review by archbishop12 — APPROVED** [review]
@moseskamira I am merging this.
